### PR TITLE
add missing extensions for TensorFlow 1.10.1 (REVIEW)

### PR DESCRIPTION
**Pull request by boegel**
required ~~because of the changes to the `TensorFlow` easyblock, see https://github.com/easybuilders/easybuild-easyblocks/pull/1537~~ to fix problems reported in https://github.com/easybuilders/easybuild-easyblocks/issues/1536

cc @akesandgren, @agijsberts

### CODE REVIEW

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 3 out of 3 (3 easyconfigs in this PR)
node2419.golett.os - Linux centos linux 7.5.1804, Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50GHz, Python 2.7.5
See https://gist.github.com/b683b245dd94f27882f962b2c31e7df4 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 3 out of 3 (3 easyconfigs in this PR)
node3170.skitty.os - Linux centos linux 7.5.1804, Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz, Python 2.7.5
See https://gist.github.com/692e844e0bea59d97d34802772974d3b for a full test report.

**Comment by akesandgren** [comment]
Test report by @akesandgren
**FAILED**
Build succeeded for 1 out of 3 (1 easyconfigs in this PR)
b-an03.hpc2n.umu.se - Linux ubuntu 16.04, Intel(R) Xeon(R) CPU E5-2690 v4 @ 2.60GHz, Python 2.7.12
See https://gist.github.com/1c20b59ee6ae149518ad59db78576b03 for a full test report.

**Comment by agijsberts** [comment]
Unless I'm mistaking, TensorFlow's build seems to use it's own external protobuf with no straightforward way (yet) to use system's protobuf:
https://github.com/tensorflow/tensorflow/issues/16104

To avoid interoperability issues (I've been bitten by them multiple times), you are probably better off using protobuf 3.6.0, which is the same as TensorFlow's 1.10 packaged version:
https://github.com/tensorflow/tensorflow/blob/r1.10/tensorflow/contrib/cmake/external/protobuf.cmake

Alternatively, you could of course also patch the tag in this cmake file to 3.6.1.

File: easybuild/easyconfigs/p/protobuf/protobuf-3.6.1-foss-2018b.eb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+name = 'protobuf'
+version = '3.6.1'
+
+homepage = 'https://github.com/google/protobuf/'
+description = """Google Protocol Buffers"""
+
+toolchain = {'name': 'foss', 'version': '2018b'}
+
+source_urls = ['https://github.com/google/protobuf/releases/download/v%(version)s/']
+sources = ['protobuf-cpp-%(version)s.tar.gz']
```
Comment (akesandgren) [inline easybuild/easyconfigs/p/protobuf/protobuf-3.6.1-foss-2018b.eb]: Why the change to protobuf-cpp?
Older protobuf versions (currently in develop) uses protobuf-3.x.y.tar.gz

File: easybuild/easyconfigs/p/protobuf/protobuf-3.6.1-fosscuda-2018b.eb [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+name = 'protobuf'
+version = '3.6.1'
+
+homepage = 'https://github.com/google/protobuf/'
+description = """Google Protocol Buffers"""
+
+toolchain = {'name': 'fosscuda', 'version': '2018b'}
+
+source_urls = ['https://github.com/google/protobuf/releases/download/v%(version)s/']
+sources = ['protobuf-cpp-%(version)s.tar.gz']
```
Comment (akesandgren) [inline easybuild/easyconfigs/p/protobuf/protobuf-3.6.1-fosscuda-2018b.eb]: And of course, same here.

**Comment by akesandgren** [comment]
Test report by @akesandgren
**FAILED**
Build succeeded for 1 out of 2 (1 easyconfigs in this PR)
b-an03.hpc2n.umu.se - Linux ubuntu 16.04, Intel(R) Xeon(R) CPU E5-2690 v4 @ 2.60GHz, Python 2.7.12
See https://gist.github.com/38d466d20bddd4bf3a84c38201d18539 for a full test report.

File: easybuild/easyconfigs/t/TensorFlow/TensorFlow-1.10.0-fosscuda-2018b-Python-2.7.15.eb [hunk 3]
```diff
@@ -31,9 +31,56 @@ builddependencies = [
...
+        'checksums': ['15a5fc23ba2e5f30758fbb8763f79a764e3de1ffcb92c2ccd46f45ffa63910d2'],
+        'modulename': 'grpc',
+    }),
+    ('Markdown', '3.0.1', {
+        'source_urls': ['https://pypi.python.org/packages/source/M/Markdown'],
+        'checksums': ['d02e0f9b04c500cde6637c11ad7c72671f359b87b9fe924b2383649d8841db7c'],
+    }),
+    ('tensorboard', '1.10.0', {
+        'source_urls': ['https://pypi.python.org/packages/source/t/tensorboard'],
+        'source_tmpl': 'tensorboard-%(version)s-py3-none-any.whl',
```
Comment (akesandgren) [inline easybuild/easyconfigs/t/TensorFlow/TensorFlow-1.10.0-fosscuda-2018b-Python-2.7.15.eb]: Should be -py2- here.

File: easybuild/easyconfigs/t/TensorFlow/TensorFlow-1.10.1-fosscuda-2018b-Python-2.7.15.eb [hunk 4]
```diff
@@ -31,9 +31,56 @@ builddependencies = [
...
+        'checksums': ['15a5fc23ba2e5f30758fbb8763f79a764e3de1ffcb92c2ccd46f45ffa63910d2'],
+        'modulename': 'grpc',
+    }),
+    ('Markdown', '3.0.1', {
+        'source_urls': ['https://pypi.python.org/packages/source/M/Markdown'],
+        'checksums': ['d02e0f9b04c500cde6637c11ad7c72671f359b87b9fe924b2383649d8841db7c'],
+    }),
+    ('tensorboard', '1.10.0', {
+        'source_urls': ['https://pypi.python.org/packages/source/t/tensorboard'],
+        'source_tmpl': 'tensorboard-%(version)s-py3-none-any.whl',
```
Comment (akesandgren) [inline easybuild/easyconfigs/t/TensorFlow/TensorFlow-1.10.1-fosscuda-2018b-Python-2.7.15.eb]: and also -py2- here

**Comment by akesandgren** [comment]
Test report by @akesandgren
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in this PR)
b-an03.hpc2n.umu.se - Linux ubuntu 16.04, Intel(R) Xeon(R) CPU E5-2690 v4 @ 2.60GHz, Python 2.7.12
See https://gist.github.com/b70e6cca61e1bf935e5f2794f1c05284 for a full test report.

File: easybuild/easyconfigs/t/TensorFlow/TensorFlow-1.10.0-fosscuda-2018b-Python-2.7.15.eb [hunk 5]
```diff
@@ -31,9 +31,56 @@ builddependencies = [
...
+        'modulename': 'grpc',
+    }),
+    ('Markdown', '3.0.1', {
+        'source_urls': ['https://pypi.python.org/packages/source/M/Markdown'],
+        'checksums': ['d02e0f9b04c500cde6637c11ad7c72671f359b87b9fe924b2383649d8841db7c'],
+    }),
+    ('tensorboard', '1.10.0', {
+        'source_urls': ['https://pypi.python.org/packages/source/t/tensorboard'],
+        'source_tmpl': 'tensorboard-%(version)s-py2-none-any.whl',
+        'checksums': ['e4ea6ac2e47bf715b915f08a186e6205fa097318bd73f0b265d437b1d7834484'],
```
Comment (akesandgren) [inline easybuild/easyconfigs/t/TensorFlow/TensorFlow-1.10.0-fosscuda-2018b-Python-2.7.15.eb]: You forgot to change the checksum :-)

File: easybuild/easyconfigs/t/TensorFlow/TensorFlow-1.10.1-fosscuda-2018b-Python-2.7.15.eb [hunk 6]
```diff
@@ -31,9 +31,56 @@ builddependencies = [
...
+        'modulename': 'grpc',
+    }),
+    ('Markdown', '3.0.1', {
+        'source_urls': ['https://pypi.python.org/packages/source/M/Markdown'],
+        'checksums': ['d02e0f9b04c500cde6637c11ad7c72671f359b87b9fe924b2383649d8841db7c'],
+    }),
+    ('tensorboard', '1.10.0', {
+        'source_urls': ['https://pypi.python.org/packages/source/t/tensorboard'],
+        'source_tmpl': 'tensorboard-%(version)s-py2-none-any.whl',
+        'checksums': ['e4ea6ac2e47bf715b915f08a186e6205fa097318bd73f0b265d437b1d7834484'],
```
Comment (akesandgren) [inline easybuild/easyconfigs/t/TensorFlow/TensorFlow-1.10.1-fosscuda-2018b-Python-2.7.15.eb]: Same her

**Comment by boegel** [comment]
> Unless I'm mistaking, TensorFlow's build seems to use it's own external protobuf with no straightforward way (yet) to use system's protobuf:
> [tensorflow/tensorflow#16104](https://github.com/tensorflow/tensorflow/issues/16104)
> 
> To avoid interoperability issues (I've been bitten by them multiple times), you are probably better off using protobuf 3.6.0, which is the same as TensorFlow's 1.10 packaged version:
> https://github.com/tensorflow/tensorflow/blob/r1.10/tensorflow/contrib/cmake/external/protobuf.cmake
> 
> Alternatively, you could of course also patch the tag in this cmake file to 3.6.1.

@agijsberts I understand how this can be a problem if the `protobuf` version is significantly different, but here it's just a (very minor, based on the 3.6.1 release notes) bugfix difference, i.e. both the 'major' (`3`) and 'minor' (`6`) versions match, so it shouldn't be a problem w.r.t. mismatching symbols?

**Comment by agijsberts** [comment]
@boegel Unfortunately the situation is more tricky than it seems, as I noticed when I thought to combine ``TensorFlow 1.11`` with ``DALI 0.3.0``, specifying (python-)protobuf 3.6.1 in both of them. In short, the problem is that ``TensorFlow`` will ship with its own ``protobuf`` headers, which are independent of the ``protobuf`` version you specify in your EasyConfig.

In more detail: regardless of the EasyBuild dependency, TensorFlow will build with its internal protobuf 3.6.0. It also installs the header files for these in ``lib/python3.6/site-packages/tensorflow/include/google/protobuf/``. It is important to note that the ``google/protobuf/stubs/common.h`` header defines ``GOOGLE_PROTOBUF_VERSION`` and ``GOOGLE_PROTOBUF_MIN_PROTOC_VERSION`` (e.g., ``3006000`` or ``3006001``).

The problem stats when a module (like ``DALI``) depends both on ``TensorFlow`` as well as ``protobuf``. It will use ``protoc`` from the ``protobuf`` dependency, but it will use the protobuf headers from either ``TensorFlow`` or ``protobuf`` package, depending on the order in which the modules are loaded! Suppose our DALI uses ``protoc`` 3.6.1, it will generates files that require ``GOOGLE_PROTOBUF_VERSION`` at least ``3006001``. But if the TensorFlow includes took precedence (it will depend on the order), then ``GOOGLE_PROTOBUF_VERSION`` will be fixed at its internal version ``3006000``. Due to this version mismatch, one receives

```
error: #error This file was generated by a newer version of protoc which is
error: #error incompatible with your Protocol Buffer headers. Please update
error: #error your headers.
```

This error is confusing, since both modules may have specified the same version 3.6.1. The real problem is of course that TensorFlow forces it's own dependencies, but at the moment there doesn't seem a straightforward way to tell it to instead use the system's packages.

**Comment by akesandgren** [comment]
Test report by @akesandgren
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
b-an03.hpc2n.umu.se - Linux ubuntu 16.04, Intel(R) Xeon(R) CPU E5-2690 v4 @ 2.60GHz, Python 2.7.12
See https://gist.github.com/abec19a8f96918dceefee4f2d17cc6f4 for a full test report.

**Comment by boegel** [comment]
@agijsberts Did you verify is the problem can be circumvented completely by sticking to the same `protobuf` version (EB-dependency-wise) as the one that is "bundled" with `TensorFlow`?

Any idea if just installing the `protobuf-python` Python bindings as an extension in the TensorFlow easyconfig file (effectively reusing the `protobuf` bundled with `TensorFlow` rather than forcing our own) is an option?

BTW, there was a discussion on allowing to use a system `protobuf` recently in the TF SIG Build, see https://groups.google.com/a/tensorflow.org/forum/#!topic/build/sSOdm79w-DI

**Comment by boegel** [comment]
@agijsberts Did you verify is the problem can be circumvented completely by sticking to the same `protobuf` version (EB-dependency-wise) as the one that is "bundled" with `TensorFlow`?

Any idea if just installing the `protobuf-python` Python bindings as an extension in the TensorFlow easyconfig file (effectively reusing the `protobuf` bundled with `TensorFlow` rather than forcing our own) is an option?

BTW, there was a discussion on allowing to use a system `protobuf` recently in the TF SIG Build, see https://groups.google.com/a/tensorflow.org/forum/#!topic/build/sSOdm79w-DI

**Comment by agijsberts** [comment]
@boegel I'm not sure if I explained myself well. The issue is basically that it's very easy ending up mixing the ``protoc 3.6.1`` executable with the ``3.6.0`` headers packaged internally in ``TensorFlow``. If one simply uses ``protobuf 3.6.0`` in ``TensorFlow`` and modules that interface with it, then things work because the versions will be the same anyways. It's not a perfect solution, but I have confirmed locally that it works with a ``DALI 0.3.0`` build.

I'm not quite sure if I understand your suggestion regarding ``protobuf-python``, but mind that it's TensorFlow's wheel package itself that installs these headers in ``lib/python3.6/site-packages/tensorflow/include`` (actually, it seems to do the same for ``Eigen``). Any way to get rid of the ``protobuf`` headers in this location would probably avoid the particular problem I'm warning against, but _might_ create other problems.

**Comment by akesandgren** [comment]
To knit pick a bit, TF does not include these headers in its source distribution, it downloads them during the build process, a download that could be patched to pick 3.6.1 instead of 3.6.0, but that's another matter.

It does contain wrapper headers that includes the protobuf headers, but not protobuf itself.

**Comment by boegel** [comment]
@agijsberts This is what I meant w.r.t. `protobuf-python`: only include the Python bindings as an extension, don't include `protobuf` itself as a dependency:

```diff
diff --git a/easybuild/easyconfigs/t/TensorFlow/TensorFlow-1.10.1-foss-2018b-Python-3.6.6.eb b/easybuild/easyconfigs/t/TensorFlow/TensorFlow-1.10.1-foss-2018b-Python-3.6.6.eb
index f814c3ba6..a7d0e38f8 100644
--- a/easybuild/easyconfigs/t/TensorFlow/TensorFlow-1.10.1-foss-2018b-Python-3.6.6.eb
+++ b/easybuild/easyconfigs/t/TensorFlow/TensorFlow-1.10.1-foss-2018b-Python-3.6.6.eb
@@ -26,10 +26,7 @@ builddependencies = [
     ('wheel', '0.31.1', versionsuffix),
 ]
 
-dependencies = [
-    ('Python', '3.6.6'),
-    ('protobuf-python', '3.6.1', versionsuffix),
-]
+dependencies = [('Python', '3.6.6')]
 
 exts_defaultclass = 'PythonPackage'
 exts_default_options = {
@@ -39,6 +36,10 @@ exts_default_options = {
 exts_filter = ("python -c 'import %(ext_name)s'", '')
 
 exts_list = [
+    ('protobuf', '3.6.0', {
+        'source_urls': ['https://pypi.python.org/packages/source/p/protobuf'],
+        'modulename': 'google.protobuf',
+    }),
     ('absl-py', '0.5.0', {
         'source_urls': ['https://pypi.python.org/packages/source/a/absl-py'],
         'checksums': ['6fcc3c04dc881fd93d793674a42ee8c73155570eda8f8b90c4477c8522478b7b'],
```

That prevents having a different `protoc` all together, at least for `TensorFlow` itself it does...

I'm not entirely sure this is a good option though, since if (a different version of) `protobuf` is included as a dependency elsewhere (e.g. for `DALI`), you would still be in trouble.

So, I'll go forward with including `protobuf-python` `3.6.0` as a dependency instead, which should work for `TensorFlow` 1.10.x.
 That may not work for other versions though (both older & future versions), since based on that SIG Build thread I mentioned earlier `TensorFlow` has been using unreleased versions of `protobuf` internally, so it may not be possible to specify a particular version as a dependency that's fully compatible...

That also means that applying these same fixes to older `TensorFlow` easyconfigs is going to be a huge PITA, and probably not worth the effort (which means https://github.com/easybuilders/easybuild-easyblocks/pull/1537 will need some further tweaks).

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 2 out of 3 (1 easyconfigs in this PR)
node2406.golett.os - Linux centos linux 7.5.1804, Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50GHz, Python 2.7.5
See https://gist.github.com/af3639c52424352699df0ecbb1705a04 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
node3101.skitty.os - Linux centos linux 7.5.1804, Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz, Python 2.7.5
See https://gist.github.com/bd2c2bd2e8a35b804ae0f3be81f60c81 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 2 out of 3 (1 easyconfigs in this PR)
node2001.delcatty.os - Linux centos linux 7.5.1804, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.7.5
See https://gist.github.com/77ccbf55a2b0714a6f6fce0ccd7fb0f4 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
node2486.golett.os - Linux centos linux 7.5.1804, Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50GHz, Python 2.7.5
See https://gist.github.com/2995e6fce3f466a66d1736a913a54b86 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 0 out of 2 (2 easyconfigs in this PR)
nic169 - Linux centos linux 7.5.1804, Intel(R) Xeon(R) CPU E5-2650 v4 @ 2.20GHz, Python 2.7.5
See https://gist.github.com/3c93b34b7d60048e9f1b92ce0df35308 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
node2120.delcatty.os - Linux centos linux 7.5.1804, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.7.5
See https://gist.github.com/6131f86807009329a9b85cd00b1d9438 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 2 out of 2 (2 easyconfigs in this PR)
nic169 - Linux centos linux 7.5.1804, Intel(R) Xeon(R) CPU E5-2650 v4 @ 2.20GHz, Python 2.7.5
See https://gist.github.com/73f14d923aaba3b8a3847f458d93bb4e for a full test report.
